### PR TITLE
New Architectures for ST

### PR DESCRIPTION
**Pull request by brianyan918**
This PR primarily modifies espnet2/st/espnet_model.py and espnet2/bin/st_inference.py to support more architectures:
- [x] multi-decoder
- [x] ST CTC
- [x] RNN-T

Results:

### CODE REVIEW

File: egs2/covost2/asr1/local/path.sh [hunk 1]
```diff
@@ -1,8 +0,0 @@
-# This is especially useful for ST and MT
```
Comment (ftshijt) [inline egs2/covost2/asr1/local/path.sh:1]: Is there specific reason that we need to remove the check?

File: egs2/covost2/st1/conf/slurm.conf [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
 option num_threads=1 --cpus-per-task 1
 option num_nodes=* --nodes $0
 default gpu=0
-option gpu=0 -p cpu
-option gpu=* -p gpu --gres=gpu:$0 -c $0  # Recommend allocating more CPU than, or equal to the number of GPU
+option gpu=0 -p cpu --mem-per-cpu 2000M --account=bbjs-delta-cpu
+option gpu=1 -p gpuA100x4 --gres=gpu:1 -c 16 --mem 60000M --account=bbjs-delta-gpu
+option gpu=4 -p gpuA40x4 --gres=gpu:4 -c 64 --mem 240000M --account=bbjs-delta-gpu
+option gpu=* -p gpuA40x4 --gres=gpu:$0 -c 32 --mem 120000M --account=bbjs-delta-gpu
+# Recommend allocating more CPU than, or equal to the number of GPU
```
Comment (ftshijt) [inline egs2/covost2/st1/conf/slurm.conf:16]:
```suggestion
option gpu=0 -p cpu
option gpu=* -p gpu --gres=gpu:$0 -c $0  # Recommend allocating more CPU than, or equal to the number of GPU
```

File: egs2/covost2/st1/conf/slurm.conf [hunk 3]
```diff
@@ -1,14 +1,18 @@
 # Default configuration
 command sbatch --export=PATH
 option name=* --job-name $0
+default time=2-00:00:00
```
Comment (ftshijt) [inline egs2/covost2/st1/conf/slurm.conf:4]:
```suggestion
```

File: egs2/covost2/st1/run.sh [hunk 4]
```diff
@@ -1,12 +1,12 @@
 #!/usr/bin/env bash
 # Set bash to 'debug' mode, it will exit on :
 # -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
-set -e
-set -u
-set -o pipefail
+#set -e
+#set -u
+#set -o pipefail
```
Comment (ftshijt) [inline egs2/covost2/st1/run.sh:6]:
```suggestion
set -e
set -u
set -o pipefail
```

File: egs2/fisher_callhome_spanish/st1/conf/slurm.conf [hunk 5]
```diff
@@ -1,14 +1,18 @@
 # Default configuration
 command sbatch --export=PATH
 option name=* --job-name $0
+default time=2-00:00:00
```
Comment (ftshijt) [inline egs2/fisher_callhome_spanish/st1/conf/slurm.conf:4]: Ditto

File: egs2/fisher_callhome_spanish/st1/run.sh [hunk 6]
```diff
@@ -1,9 +1,9 @@
 #!/usr/bin/env bash
 # Set bash to 'debug' mode, it will exit on :
 # -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
-set -e
-set -u
-set -o pipefail
+#set -e
```
Comment (ftshijt) [inline egs2/fisher_callhome_spanish/st1/run.sh:4]: ditto

File: egs2/must_c_v2/asr1/cmd.sh [hunk 7]
```diff
@@ -28,7 +28,7 @@
 
 
 # Select the backend used by run.sh from "local", "stdout", "sge", "slurm", or "ssh"
-cmd_backend='local'
+cmd_backend='slurm'
```
Comment (ftshijt) [inline egs2/must_c_v2/asr1/cmd.sh:31]: ditto

File: egs2/TEMPLATE/st1/st.sh [hunk 8]
```diff
@@ -1330,8 +1356,6 @@ if ! "${skip_train}"; then
                 --valid_shape_file "${st_stats_dir}/valid/speech_shape" \
                 --valid_shape_file "${st_stats_dir}/valid/text_shape.${tgt_token_type}" \
                 --resume true \
-                --init_param ${pretrained_asr} \
-                --ignore_init_mismatch ${ignore_init_mismatch} \
```
[the remark this answers is not in the document]
Comment (brianyan918) [reply egs2/TEMPLATE/st1/st.sh:1334]: I think it's better to allow these to be input from the config file. What do you think?

[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/TEMPLATE/st1/st.sh:1334]: If you intend to remove, please also remove the related arguments in help_msg above and clarify that a bit in the coming documentation~

**Comment by ftshijt** [comment]
Is it ready to merge?

**Comment by sw005320** [comment]
Is it ready for me to review it?

**Comment by brianyan918** [comment]
> Is it ready for me to review it?

Would you mind starting to review it? I changed a lot of files

File: egs2/TEMPLATE/st1/local/images/asr-init.png
Comment (brianyan918) [inline egs2/TEMPLATE/st1/local/images/asr-init.png:1]: @sw005320 I added some images for my README document. Where would be the best place for this?

File: espnet2/asr/transducer/beam_search_transducer_streaming.py [hunk 9]
```diff
@@ -0,0 +1,1265 @@
+"""Search algorithms for Transducer models."""
```
Comment (b-flo) [inline espnet2/asr/transducer/beam_search_transducer_streaming.py:1]: Hi @brianyan918 ! Not a review but I was wondering if you plan to adapt `beam_search_transducer.py` for streaming purposes instead? As far as I saw, the difference seem minors.

Also, what's the (future or current) purpose of `beam_search_default_v2`?

File: espnet/nets/beam_search_timesync_streaming.py [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+        self.pre_beam_size = int(pre_beam_ratio * beam_size)
+        self.ctc_weight = weights["ctc"]
+        self.lm_weight = weights["lm"]
+        self.decoder_weight = weights["decoder"]
+        self.penalty = weights["length_bonus"]
+        self.blank_penalty = (
+            np.log(weights["blank_penalty"]) if "blank_penalty" in weights else 0.0
+        )
+        # self.sos = sos
+        self.sos = 0  # hacked
```
Comment (brianyan918) [inline espnet/nets/beam_search_timesync_streaming.py:73]: fix this

File: espnet/nets/beam_search_timesync_streaming.py [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+            decoder_scores, decoder_state = self.decoder.score(
+                self.sos_th, init_decoder_state, enc_output
+            )
+            self.attn_cache[(self.sos,)] = CacheItem(
+                state=decoder_state,
+                scores=decoder_scores,
+                log_sum=0.0,
+            )
+
+            # hacked
```
Comment (brianyan918) [inline espnet/nets/beam_search_timesync_streaming.py:100]: fix this

File: espnet/nets/beam_search_timesync_streaming.py [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+        self,
+        h: Tuple[int],
+        cache: dict,
+        scorer: ScorerInterface,
+        recompute_cache: bool = False,
+    ) -> Any:
+        """Retrieve decoder/LM scores which may be cached."""
+        root = h[:-1]  # prefix
+
+        # if (root in cache and not recompute_cache) or len(root) == 1:
```
Comment (brianyan918) [inline espnet/nets/beam_search_timesync_streaming.py:130]: fix this

File: espnet/nets/batch_beam_search.py [hunk 13]
```diff
@@ -159,11 +180,33 @@ def score_full(
         scores = dict()
         states = dict()
         for k, d in self.full_scorers.items():
-            scores[k], states[k] = d.batch_score(hyp.yseq, hyp.states[k], x)
+            if "decoder" in k and self.return_hs:
+                scores[k], hs, states[k] = d.batch_score(
+                    hyp.yseq, hyp.states[k], x, return_hs=self.return_hs
+                )
+            elif "decoder" in k and pre_x is not None:
+                if pre_x2 is not None:
```
Comment (brianyan918) [inline espnet/nets/batch_beam_search.py:188]: remove pre_x2

File: espnet/nets/batch_beam_search_online.py [hunk 14]
```diff
@@ -128,53 +166,204 @@ def forward(
...
+                if "decoder" in self.scorers.keys():
+                    init_scores["decoder"] = torch.tensor([0.0])
+                    init_states["decoder"] = [None]
+                self.running_hyps = BatchHypothesis(
+                    score=torch.tensor([0.0]),
+                    scores=init_scores,
+                    states=init_states,
+                    length=torch.tensor([2]),
+                    yseq=torch.tensor(
+                        [[self.sos, 250003]], device=x.device
```
Comment (brianyan918) [inline espnet/nets/batch_beam_search_online.py:198]: fix this

File: espnet/nets/batch_beam_search_online.py
Comment (sw005320) [inline espnet/nets/batch_beam_search_online.py:1]: I want @ftshijt to check this part carefully.

File: egs2/must_c_v2/st1/README.md
Comment (sw005320) [inline egs2/must_c_v2/st1/README.md:1]: Add config, checkpoint link, and other information (later?).

File: espnet/nets/beam_search.py [hunk 15]
```diff
@@ -17,6 +17,7 @@ class Hypothesis(NamedTuple):
     score: Union[float, torch.Tensor] = 0
     scores: Dict[str, Union[float, torch.Tensor]] = dict()
     states: Dict[str, Any] = dict()
+    hs: List[torch.Tensor] = []
```
Comment (sw005320) [inline espnet/nets/beam_search.py:20]: ditto.
Please explain this.

File: espnet/nets/beam_search.py [hunk 16]
```diff
@@ -289,13 +303,17 @@ def merge_states(self, states: Any, part_states: Any, part_idx: int) -> Any:
...
+        running_hyps: List[Hypothesis],
+        x: torch.Tensor,
+        pre_x: torch.Tensor = None,
     ) -> List[Hypothesis]:
         """Search new tokens for running hypotheses and encoded speech x.
 
         Args:
             running_hyps (List[Hypothesis]): Running hypotheses on beam
             x (torch.Tensor): Encoded speech feature (T, D)
+            pre_x (torch.Tensor): Encoded speech feature for sequential attn (T, D)
```
Comment (sw005320) [inline espnet/nets/beam_search.py:316]: What does "for sequential attn" mean?
can you elaborate on the information?

File: espnet/nets/beam_search_timesync_streaming.py [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+        score = root_log_sum + cand_score
+
+        logging.debug("cand score: " + str(cand_score))
+        logging.debug("decoder score: " + str(score))
+
+        self.block_set.add(root)
+
+        return score
+
+    def joint_score(
```
Comment (sw005320) [inline espnet/nets/beam_search_timesync_streaming.py:162]: Maybe better to make this part generalized (can join multiple scores with some weights like label-sync beam search) in the future.
You can leave some comments (TODO?)

File: espnet/nets/pytorch_backend/transformer/decoder_layer.py [hunk 18]
```diff
@@ -43,24 +43,39 @@ def __init__(
         dropout_rate,
         normalize_before=True,
         concat_after=False,
+        sequential_attn=None,
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/transformer/decoder_layer.py:46]: What is sequential attention?
can you add an explanation?

File: espnet2/asr/ctc.py [hunk 19]
```diff
@@ -62,6 +68,59 @@ def loss_fn(self, th_pred, th_target, th_ilen, th_olen) -> torch.Tensor:
                 loss = loss / size
             return loss
 
+        elif self.ctc_type == "builtin2":
```
Comment (sw005320) [inline espnet2/asr/ctc.py:71]: can you add some clarification on builtin vs. builtin2?

File: espnet2/asr/decoder/transformer_decoder.py
Comment (sw005320) [inline espnet2/asr/decoder/transformer_decoder.py:1]: ditto.
It is better to add en explanation about `hs`

File: espnet2/asr/transducer/beam_search_transducer_streaming.py
Comment (sw005320) [inline espnet2/asr/transducer/beam_search_transducer_streaming.py:1]: I will check it after the unit test is added.
Please ping me then.

File: espnet2/text/hugging_face_tokenizer.py [hunk 20]
```diff
@@ -12,6 +12,10 @@
 except ImportError:
     is_transformers_available = False
 
+import os
+
+os.environ["CURL_CA_BUNDLE"] = ""
```
Comment (sw005320) [inline espnet2/text/hugging_face_tokenizer.py:17]: what's this?

**Review by sw005320 — COMMENTED** [review]
I leave some comments.
I think it looks good shape to me.
Can you start to add some tests?

File: espnet2/text/hugging_face_tokenizer.py (hunk 20 above)
Comment (brianyan918) [reply espnet2/text/hugging_face_tokenizer.py:17]: its a useless artifact

File: espnet2/asr/ctc.py (hunk 19 above)
Comment (brianyan918) [reply espnet2/asr/ctc.py:71]: Builtin2 was the old way that espnet was handling nan in CTC. At some point, it was switched to using the zero_infinity flag in pytorch's CTC - I don't know the details of how they do that.

But I found that there are differences in the performance of resulting models. For instance, builtin2 was 0.3 bleu better on MuST-C en-de. This was the result we published in the EACL paper. I thought it might be better to keep builtin2 as an option.

I will add a description to the code.

File: egs2/must_c_v2/st1/README.md
Comment (brianyan918) [reply egs2/must_c_v2/st1/README.md:1]: working on it!

Comment (brianyan918) [reply egs2/must_c_v2/st1/README.md:1]: but maybe will finish it in a later pr

File: egs2/must_c_v2/st1/conf/tuning/train_st_ctc_md_conformer_asrinit_v3_noamp_batch50m_ctcsamp0.1.yaml [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+    time_mask_width_ratio_range:
+    - 0.
+    - 0.05
+    num_time_mask: 5
+
+ctc_sample_rate: 0.1
+
+ignore_init_mismatch: true
+init_param:
+    - "/exp/byan/working/espnet-ml/egs2/must_c_v2/asr1/exp/asr_train_asr_conformer_raw_bpe4000_sp/valid.acc.ave_10best.pth:encoder:encoder"
```
Comment (sw005320) [inline egs2/must_c_v2/st1/conf/tuning/train_st_ctc_md_conformer_asrinit_v3_noamp_batch50m_ctcsamp0.1.yaml:119]: Is it OK to add some absolute paths?
Is this an example and the users can change it depending on their environment?
If so, can you add some comments?

**Comment by sw005320** [comment]
I leave some comments.
I think it is a very good shape now.
Please check my comments.

**Review by ftshijt — APPROVED** [review]
The whole setting looks great to me! As I mentioned in our discussion, it would be great if we can accelerate the PR (e.g., some model links/readme) might be finished in another PR if possible) as some of my s2st part will depend on the implementation. Great thanks!

**Comment by brianyan918** [comment]
@sw005320 oh thanks, i meant to change to rel path

**Comment by brianyan918** [comment]
@sw005320 is it ok to merge this?

**Comment by sw005320** [comment]
Hi @brianyan918
It would be good to finish this PR as soon as possible.
Could you fix the CI error and finish this PR?

**Comment by brianyan918** [comment]
> Hi @brianyan918
> It would be good to finish this PR as soon as possible.
> Could you fix the CI error and finish this PR?

Ok! Will do it in the evening today

**Comment by sw005320** [comment]
@brianyan918, I'd recommend you run some tests locally.
https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md

**Comment by brianyan918** [comment]
> @brianyan918, I'd recommend you run some tests locally. https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md

I had a naming conflict. Locally, the individual test worked. But when CI ran all the tests, this issue popped up.

**Comment by sw005320** [comment]
I see.
